File: artifacts/api-server/src/lib/excel/writer.ts
import type { MappingResult } from "../ai/mapper.js";
import type { CanonicalField } from "../canonical/schemas.js";

export interface ExcelRow {
  smrConstruct: string;
  compositeMappingReference: string;
  dialectTerm: string;
  canonicalXpath: string;
  erpJsonPath: string;
}

export function buildExcelRows(
  mappings: MappingResult[],
  canonicalFields: CanonicalField[]
): ExcelRow[] {
  const rows: ExcelRow[] = [];

  // Group by section to produce Record rows
  const sections = new Map<string, CanonicalField[]>();
  for (const cf of canonicalFields) {
    if (cf.smrConstruct === "Record") {
      if (!sections.has(cf.section)) {
        sections.set(cf.section, []);
      }
    }
  }

  // Add record rows for sections
  const recordFields = canonicalFields.filter(cf => cf.smrConstruct === "Record");
  for (const rf of recordFields) {
    rows.push({
      smrConstruct: "Record",
      compositeMappingReference: "",
      dialectTerm: "",
      canonicalXpath: rf.xpath,
      erpJsonPath: "",
    });
  }

  // Add composite rows
  const compositeFields = canonicalFields.filter(cf => cf.smrConstruct === "Composite");
  for (const cf of compositeFields) {
    // Find all ERP fields that mapped to children of this composite
    const compositeRef = cf.compositeRef ?? "";
    const childXpathPrefix = cf.xpath.replace(/\[.*?\]$/, "");
    const mappedChildren = mappings.filter(m =>
      m.canonicalXpath.startsWith(childXpathPrefix) && m.erpJsonPath
    );
    const erpPaths = mappedChildren.map(m => m.erpJsonPath).join(", ");

    rows.push({
      smrConstruct: "Composite",
      compositeMappingReference: compositeRef,
      dialectTerm: "",
      canonicalXpath: cf.xpath,
      erpJsonPath: erpPaths,
    });
  }

  // Add field rows
  for (const mapping of mappings) {
    if (!mapping.canonicalXpath) {
      // Unmapped — add a row with empty canonical fields
      rows.push({
        smrConstruct: "Field",
        compositeMappingReference: "",
        dialectTerm: "",
        canonicalXpath: "",
        erpJsonPath: mapping.erpJsonPath,
      });
      continue;
    }
    rows.push({
      smrConstruct: mapping.smrConstruct,
      compositeMappingReference: mapping.compositeRef,
      dialectTerm: mapping.dialectTerm,
      canonicalXpath: mapping.canonicalXpath,
      erpJsonPath: mapping.erpJsonPath,
    });
  }

  // Sort: Records first, then Composites, then Fields
  rows.sort((a, b) => {
    const order = { Record: 0, Composite: 1, Field: 2 };
    return (order[a.smrConstruct as keyof typeof order] ?? 2) - (order[b.smrConstruct as keyof typeof order] ?? 2);
  });

  return rows;
}

export function rowsToCsv(rows: ExcelRow[]): string {
  const headers = ["SMR Construct", "Composite Mapping Reference", "Dialect Term", "Canonical XPath", "ERP Json Path"];
  const lines = [headers.join(",")];
  for (const row of rows) {
    const values = [
      row.smrConstruct,
      row.compositeMappingReference,
      row.dialectTerm,
      row.canonicalXpath,
      row.erpJsonPath,
    ].map(v => `"${(v ?? "").replace(/"/g, '""')}"`);
    lines.push(values.join(","));
  }
  return lines.join("\n");
}

File: artifacts/api-server/lib/excel/writer.py
import csv
import io
from dataclasses import dataclass

@dataclass
class ExcelRow:
    smr_construct: str
    composite_mapping_reference: str
    dialect_term: str
    canonical_xpath: str
    erp_json_path: str

def build_excel_rows(mappings: list[dict], canonical_fields: list[dict]) -> list[ExcelRow]:
    rows: list[ExcelRow] = []

    record_fields = [cf for cf in canonical_fields if cf["smr_construct"] == "Record"]
    for rf in record_fields:
        rows.append(ExcelRow(
            smr_construct="Record",
            composite_mapping_reference="",
            dialect_term="",
            canonical_xpath=rf["xpath"],
            erp_json_path="",
        ))

    composite_fields = [cf for cf in canonical_fields if cf["smr_construct"] == "Composite"]
    for cf in composite_fields:
        child_prefix = cf["xpath"].split("[")[0] if "[" in cf["xpath"] else cf["xpath"]
        mapped_children = [m for m in mappings if m.get("canonicalXpath", "").startswith(child_prefix) and m.get("erpJsonPath")]
        erp_paths = ", ".join(m["erpJsonPath"] for m in mapped_children)
        rows.append(ExcelRow(
            smr_construct="Composite",
            composite_mapping_reference=cf.get("composite_ref") or "",
            dialect_term="",
            canonical_xpath=cf["xpath"],
            erp_json_path=erp_paths,
        ))

    for mapping in mappings:
        if not mapping.get("canonicalXpath"):
            rows.append(ExcelRow(
                smr_construct="Field",
                composite_mapping_reference="",
                dialect_term="",
                canonical_xpath="",
                erp_json_path=mapping.get("erpJsonPath", ""),
            ))
        else:
            rows.append(ExcelRow(
                smr_construct=mapping.get("smrConstruct", "Field"),
                composite_mapping_reference=mapping.get("compositeRef", ""),
                dialect_term=mapping.get("dialectTerm", ""),
                canonical_xpath=mapping.get("canonicalXpath", ""),
                erp_json_path=mapping.get("erpJsonPath", ""),
            ))

    order = {"Record": 0, "Composite": 1, "Field": 2}
    rows.sort(key=lambda r: order.get(r.smr_construct, 2))
    return rows

def rows_to_csv(rows: list[ExcelRow]) -> str:
    output = io.StringIO()
    writer = csv.writer(output, quoting=csv.QUOTE_ALL)
    writer.writerow(["SMR Construct", "Composite Mapping Reference", "Dialect Term", "Canonical XPath", "ERP Json Path"])
    for row in rows:
        writer.writerow([
            row.smr_construct,
            row.composite_mapping_reference,
            row.dialect_term,
            row.canonical_xpath,
            row.erp_json_path,
        ])
    return output.getvalue()
